### PR TITLE
Create organizations resource

### PR DESCRIPTION
**Pull request by zoeblairfriedman**
### This PR creates the backend entity for Organizations. This includes:
 
DTOs to create and update
Organization Entity
Organizations Controller
Organizations Module
Organization Service

### It accomplishes:

**1. Make http POST request to create new organization**
Postman post request: 
post_org

Proof that it exists in the database:
post_org_db

Code snippet for testing a POST request to ```http://localhost:3001/api/organizations``` 
```
{
    "name": "Very Good Donators",
    "description": "We donate things",
    "website": "www.gooddonators.com",
    "phone": "555-555-5555",
    "city": "Newark",
    "state": "New Jersey",
    "ein": 123456789,
    "tax_exempt_id": 987654321,
    "address": "123 Olde Springs Rd"
}
```
Add a second organization:
```
{
    "name": "ABC Volunteers",
    "description": "We are volunteers",
    "website": "www.abcvolunteers.com",
    "phone": "555-666-7777",
    "city": "Madison",
    "state": "Wisconsin",
    "ein": 999999999,
    "tax_exempt_id": 88888888,
    "address": "55 Best Street"
}
```

**2. Make http GET request to retrieve all organizations**
You should now see the two organizations we just created:
get_all_organizations

**3. Make http GET request to retrieve organizations by id**
Make your get request to ```http://localhost:3001/api/organizations/<organization-id-previously-created>``` to retrieve the first org we created: 
get_one_organization

**4. Make http PATCH request to update organization**

DB before:
update_before

Postman patch request: 
patch_org

DB after:
update_after

Code snippet to test in your patch request to ``` http://localhost:3001/api/organizations/<organization-id-previously-created>``` 
```
{
    "name": "Very Good Donators",
    "description": "We donate things",
    "website": "www.gooddonators.com",
    "phone": "555-555-5555",
    "city": "Columbia",
    "state": "South Carolina",
    "ein": 123456789,
    "tax_exempt_id": 987654321,
    "address": "123 Olde Springs Rd"
}
```

**5. Validates uniqueness of ein**
dupe_ein

Snippet to test uniqueness of EIN validator: 
```
{
    "name": "XYZ Volunteers",
    "description": "We are also volunteers",
    "website": "www.xyzvolunteers.com",
    "phone": "555-666-7777",
    "city": "Madison",
    "state": "Wisconsin",
    "ein": 999999999,
    "tax_exempt_id": 88888888,
    "address": "55 Best Street"
}
```

### CODE REVIEW

**Comment by jd2rogers2** [comment]
@zoeblairfriedman snippets of request data would be good so reviewer/tester doesn't have to write everything out themselves!

File: server/src/app.controller.spec.ts [hunk 1]
```diff
@@ -14,6 +14,7 @@ describe('AppController', () => {
     appController = app.get<AppController>(AppController);
   });
 
+  // i'm not sure what this is:
```
Comment (jd2rogers2) [inline server/src/app.controller.spec.ts:17]: good call, our tests are woefully sad
putting in an issue now
https://github.com/Nonprofit-Exchange-Hub/web-app/pull/78
this PR will skip this pointless test

it was mocking up our app controller and then hitting an endpoint called "root" (that no longer exists) and expecting it to return that string. happy to talk through it more if still unclear

File: server/src/organizations/dto/create-organization.dto.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { Organization } from '../entities/organization.entity'
+import { PartialType } from '@nestjs/mapped-types';
+
+export class CreateOrganizationDto extends PartialType(Organization){
```
Comment (jd2rogers2) [inline server/src/organizations/dto/create-organization.dto.ts:4]: at the hackathon we talked it out and we decided we shouldn't need the `extends PartialType(Organization)` bit here
this bit says "any of the following attributes are optional" but i think for creating an organization we want to force all of these properties to be there right?

https://docs.nestjs.com/openapi/mapped-types

File: server/src/organizations/dto/update-organization.dto.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import { PartialType } from '@nestjs/mapped-types';
+import { CreateOrganizationDto } from './create-organization.dto';
+
+export class UpdateOrganizationDto extends PartialType(CreateOrganizationDto) {}
```
Comment (jd2rogers2) [inline server/src/organizations/dto/update-organization.dto.ts:4]: 👍

File: server/src/organizations/entities/organization.entity.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { Entity, Column, PrimaryGeneratedColumn } from 'typeorm';
+
+// Considering validators for uniqueness: org_name? ein? both?
```
Comment (jd2rogers2) [inline server/src/organizations/entities/organization.entity.ts:3]: this is a good question for product
ping the general or design/ux channel and maybe cc: joel so he can point you to someone who would know

File: server/src/organizations/organizations.controller.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import { CreateOrganizationDto } from './dto/create-organization.dto';
+import { UpdateOrganizationDto } from './dto/update-organization.dto';
+
+
+@Controller('organizations')
+export class OrganizationsController {
+  constructor(private readonly organizationsService: OrganizationsService) {}
+
+  @Post()
+  create(@Body() createOrganizationDto: CreateOrganizationDto) {
```
Comment (jd2rogers2) [inline server/src/organizations/organizations.controller.ts:12]: TS return types!
checkout the controllers of other resources that have been added and i think you'll be able to figure it out
i.e. [this](https://github.com/Nonprofit-Exchange-Hub/web-app/blob/19733511b87e8dcbe16b05e48530b8651f50fbb2/server/src/assets/assets.controller.ts#L44)
format is `function name(arg: argType): functionReturnType {...`

File: server/src/organizations/organizations.service.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import { UpdateOrganizationDto } from './dto/update-organization.dto';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { Organization } from './entities/organization.entity'
+
+@Injectable()
+export class OrganizationsService {
+  constructor(@InjectRepository(Organization) private organizationsRepository: Repository<Organization>) { }
+
+  async create(createOrganizationDto: CreateOrganizationDto) {
```
Comment (jd2rogers2) [inline server/src/organizations/organizations.service.ts:12]: return types for all of these too plz

File: server/src/app.controller.spec.ts (hunk 1 above)
Comment (jd2rogers2) [reply server/src/app.controller.spec.ts:17]: don't forget to remove this comment before merging

**Review by jd2rogers2 — COMMENTED** [review]
looking really good 💯 
couple small nits, and looks like we have a question for product that needs answering before we ship
and then if you happen to have the test requests saved in your postman, can you please add them in snippets to the PR description

File: server/src/organizations/entities/organization.entity.ts (hunk 4 above)
Comment (zoeblairfriedman) [reply server/src/organizations/entities/organization.entity.ts:3]: Heard back from Joel on this. He said to go with the EIN.

File: server/src/organizations/dto/create-organization.dto.ts (hunk 2 above)
Comment (zoeblairfriedman) [reply server/src/organizations/dto/create-organization.dto.ts:4]: I was wondering about this, myself. My bad, I thought we decided the opposite in the hackathon. But I agree. Will adjust!

File: server/src/app.controller.spec.ts (hunk 1 above)
Comment (zoeblairfriedman) [reply server/src/app.controller.spec.ts:17]: Lol. Sorry! I saw the TS error and meant to investigate and then forgot.  Makes sense though for sure.

File: server/src/organizations/organizations.controller.ts (hunk 5 above)
Comment (zoeblairfriedman) [reply server/src/organizations/organizations.controller.ts:12]: done!

File: server/src/organizations/organizations.service.ts (hunk 6 above)
Comment (zoeblairfriedman) [reply server/src/organizations/organizations.service.ts:12]: done!

**Comment by zoeblairfriedman** [comment]
Okay James, this should be good to go for now.

**Review by jd2rogers2 — APPROVED** [review]
this is dooooppppeee 💯 💯 💯 💯 💯 💯 
ship it!

delete request not in PR description but succeeds, here it is for posterity
```
curl --request DELETE \
--url http://localhost:3001/api/organizations/1 \
--header 'Content-Type: application/json'
```

**Pull request merged**